// @defichain/whale-api-client/dist/errors/api.error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhaleApiException = exports.WhaleApiErrorType = void 0;
var WhaleApiErrorType;
(function (WhaleApiErrorType) {
    WhaleApiErrorType["ValidationError"] = "ValidationError";
    WhaleApiErrorType["BadRequest"] = "BadRequest";
    WhaleApiErrorType["NotFound"] = "NotFound";
    WhaleApiErrorType["Conflict"] = "Conflict";
    WhaleApiErrorType["Forbidden"] = "Forbidden";
    WhaleApiErrorType["Unauthorized"] = "Unauthorized";
    WhaleApiErrorType["BadGateway"] = "BadGateway";
    WhaleApiErrorType["TimeoutError"] = "TimeoutError";
    WhaleApiErrorType["UnknownError"] = "UnknownError";
})(WhaleApiErrorType = exports.WhaleApiErrorType || (exports.WhaleApiErrorType = {}));
/**
 * Serialized exception from DeFi Whale
 */
class WhaleApiException extends Error {
    constructor(error) {
        super(`${error.code} - ${error.type} ${WhaleApiException.url(error)}${WhaleApiException.message(error)}`);
        this.error = error;
    }
    /**
     * @return {number} error code
     */
    get code() {
        return this.error.code;
    }
    /**
     * @return {string} error type
     */
    get type() {
        return this.error.type;
    }
    /**
     * @return {number} time that error occurred at
     */
    get at() {
        return this.error.at;
    }
    /**
     * @return {string} url that threw this endpoint
     */
    get url() {
        return this.error.url;
    }
    static url({ url }) {
        return url !== undefined && url !== null ? `(${url})` : '';
    }
    static message({ message }) {
        return message !== undefined && message !== null ? `: ${message}` : '';
    }
}
exports.WhaleApiException = WhaleApiException;
