// @defichain/whale-api-client/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loan = exports.fee = exports.rawtx = exports.stats = exports.prices = exports.oracles = exports.blocks = exports.masternodes = exports.tokens = exports.transactions = exports.poolpairs = exports.address = exports.rpc = void 0;
__exportStar(require("./errors"), exports);
exports.rpc = __importStar(require("./api/rpc"));
exports.address = __importStar(require("./api/address"));
exports.poolpairs = __importStar(require("./api/poolpairs"));
exports.transactions = __importStar(require("./api/transactions"));
exports.tokens = __importStar(require("./api/tokens"));
exports.masternodes = __importStar(require("./api/masternodes"));
exports.blocks = __importStar(require("./api/blocks"));
exports.oracles = __importStar(require("./api/oracles"));
exports.prices = __importStar(require("./api/prices"));
exports.stats = __importStar(require("./api/stats"));
exports.rawtx = __importStar(require("./api/rawtx"));
exports.fee = __importStar(require("./api/fee"));
exports.loan = __importStar(require("./api/loan"));
__exportStar(require("./whale.api.client"), exports);
__exportStar(require("./whale.api.response"), exports);
__exportStar(require("./whale.rpc.client"), exports);
